File: setup/associate.py


from board import Port
from board import Road
from board import Settlement
from board import Tile


def find(id: int) -> callable:
	def callback(item: object) -> bool:
		return item.id == id

	return callback


def ports_with_settlements(port_dicts: list[dict], ports: list[Port], settlements: list[Settlement]):
	for port_dict in port_dicts:
		port = next(filter(find(port_dict["id"]), ports))
		for settlement_direction, settlement_id in port_dict["Settlements"].items():
			settlement = next(filter(find(settlement_id), settlements))
			direction_index = Port.Settlements.ENUM_VALUES[settlement_direction.split("::")[2]]
			port.settlements[direction_index] = settlement


def roads_with_settlements(road_dicts: list[dict], roads: list[Road], settlements: list[Settlement]):
	for road_dict in road_dicts:
		road = next(filter(find(road_dict["id"]), roads))
		for settlement_direction, settlement_id in road_dict["Settlements"].items():
			settlement = next(filter(find(settlement_id), settlements))
			direction_index = Road.Settlements.ENUM_VALUES[settlement_direction.split("::")[2]]
			road.settlements[direction_index] = settlement


def roads_with_tiles(road_dicts: list[dict], roads: list[Road], tiles: list[Tile]):
	for road_dict in road_dicts:
		road = next(filter(find(road_dict["id"]), roads))
		for tile_direction, tile_id in road_dict["Tiles"].items():
			tile = next(filter(find(tile_id), tiles))
			direction_index = Road.Tiles.ENUM_VALUES[tile_direction.split("::")[2]]
			road.tiles[direction_index] = tile


def settlements_with_ports(settlement_dicts: list[dict], settlements: list[Settlement], ports: list[Road]):
	for settlement_dict in settlement_dicts:
		settlement = next(filter(find(settlement_dict["id"]), settlements))
		for port_direction, port_id in settlement_dict["Ports"].items():
			print(settlement.id, port_id)
			port = next(filter(find(port_id), ports))
			direction_index = Settlement.Ports.ENUM_VALUES[port_direction.split("::")[2]]
			settlement.ports[direction_index] = port


def settlements_with_roads(settlement_dicts: list[dict], settlements: list[Settlement], roads: list[Road]):
	for settlement_dict in settlement_dicts:
		settlement = next(filter(find(settlement_dict["id"]), settlements))
		for road_direction, road_id in settlement_dict["Roads"].items():
			road = next(filter(find(road_id), roads))
			direction_index = Settlement.Roads.ENUM_VALUES[road_direction.split("::")[2]]
			settlement.roads[direction_index] = road


def settlements_with_tiles(settlement_dicts: list[dict], settlements: list[Settlement], tiles: list[Settlement]):
	for settlement_dict in settlement_dicts:
		settlement = next(filter(find(settlement_dict["id"]), settlements))
		for tile_direction, tile_id in settlement_dict["Tiles"].items():
			tile = next(filter(find(tile_id), tiles))
			direction_index = Settlement.Tiles.ENUM_VALUES[tile_direction.split("::")[2]]
			settlement.tiles[direction_index] = tile


def tiles_with_roads(tile_dicts: list[dict], tiles: list[Tile], roads: list[Road]):
	for tile_dict in tile_dicts:
		tile = next(filter(find(tile_dict["id"]), tiles))
		for road_direction, road_id in tile_dict["Roads"].items():
			road = next(filter(find(road_id), roads))
			direction_index = Tile.Roads.ENUM_VALUES[road_direction.split("::")[2]]
			tile.roads[direction_index] = road


def tiles_with_settlements(tile_dicts: list[dict], tiles: list[Road], settlements: list[Settlement]):
	for tile_dict in tile_dicts:
		tile = next(filter(find(tile_dict["id"]), tiles))
		for settlement_direction, settlement_id in tile_dict["Settlements"].items():
			settlement = next(filter(find(settlement_id), settlements))
			direction_index = Tile.Settlements.ENUM_VALUES[settlement_direction.split("::")[2]]
			tile.settlements[direction_index] = settlement

File: Source/webpage/Static/Scripts/Game/Board/associate.js
export function associate_port_settlements(board_data, ports, settlements) {
    const port_dicts = board_data.ports;
    const settlement_dicts = board_data.settlements;
    for (var port_dicts_index = 0; port_dicts_index < port_dicts.length; port_dicts_index++) {
        const port_id = port_dicts[port_dicts_index].id;
        const port = ports.find(port => port.id == port_id);
        for (var port_settlement_index = 0; port_settlement_index < port.settlements.length; port_settlement_index++) {
            const settlement_id = port_dicts[port_dicts_index].settlements[port_settlement_index];
            if (settlement_id == null) {
                continue;
            }
            const settlement = settlements.find(settlement => settlement.id == settlement_id);
            port.settlements[settlement_id] = settlement;
            const settlement_dict = settlement_dicts.find(dict => dict.id == settlement_id);
            const settlement_port_index = settlement_dict.ports.indexOf(port_id);
            settlement.ports[settlement_port_index] = port;
        }
    }
}
export function associate_road_settlements(board_data, roads, settlements) {
    const road_dicts = board_data.roads;
    const settlement_dicts = board_data.settlements;
    for (var road_dicts_index = 0; road_dicts_index < road_dicts.length; road_dicts_index++) {
        const road_id = road_dicts[road_dicts_index].id;
        const road = roads.find(road => road.id == road_id);
        for (var road_settlement_index = 0; road_settlement_index < road.settlements.length; road_settlement_index++) {
            const settlement_id = road_dicts[road_dicts_index].settlements[road_settlement_index];
            if (settlement_id == null) {
                continue;
            }
            const settlement = settlements.find(settlement => settlement.id == settlement_id);
            road.settlements[settlement_id] = settlement;
            const settlement_dict = settlement_dicts.find(dict => dict.id == settlement_id);
            const settlement_road_index = settlement_dict.roads.indexOf(road_id);
            settlement.roads[settlement_road_index] = road;
        }
    }
}
export function associate_road_tiles(board_data, roads, tiles) {
    const road_dicts = board_data.roads;
    const tile_dicts = board_data.tiles;
    for (var road_dicts_index = 0; road_dicts_index < road_dicts.length; road_dicts_index++) {
        const road_id = road_dicts[road_dicts_index].id;
        const road = roads.find(road => road.id == road_id);
        for (var road_tile_index = 0; road_tile_index < road.tiles.length; road_tile_index++) {
            const tile_id = road_dicts[road_dicts_index].tiles[road_tile_index];
            if (tile_id == null) {
                continue;
            }
            const tile = tiles.find(tile => tile.id == tile_id);
            road.tiles[tile_id] = tile;
            const tile_dict = tile_dicts.find(dict => dict.id == tile_id);
            const tile_road_index = tile_dict.roads.indexOf(road_id);
            tile.roads[tile_road_index] = road;
        }
    }
}
export function associate_settlement_tiles(board_data, settlements, tiles) {
    const settlement_dicts = board_data.settlements;
    const tile_dicts = board_data.tiles;
    for (var settlement_dicts_index = 0; settlement_dicts_index < settlement_dicts.length; settlement_dicts_index++) {
        const settlement_id = settlement_dicts[settlement_dicts_index].id;
        const settlement = settlements.find(settlement => settlement.id == settlement_id);
        for (var settlement_tile_index = 0; settlement_tile_index < settlement.tiles.length; settlement_tile_index++) {
            const tile_id = settlement_dicts[settlement_dicts_index].tiles[settlement_tile_index];
            if (tile_id == null) {
                continue;
            }
            const tile = tiles.find(tile => tile.id == tile_id);
            settlement.tiles[tile_id] = tile;
            const tile_dict = tile_dicts.find(dict => dict.id == tile_id);
            const tile_settlement_index = tile_dict.settlements.indexOf(settlement_id);
            tile.settlements[tile_settlement_index] = settlement;
        }
    }
}
